### PR TITLE
Update black pre-commit hook to 21.9b0.

### PR DESCRIPTION
**Pull request by bdice**
Following up from #9300 and #9433, this PR updates the black pre-commit hook to a version that is two years newer, 21.9b0. This PR's changeset is largely the result of just running the pre-commit hook, with a couple other minor fixes. The improvements in docstring formatting are significant!

Note: I thought the stub file `column.pyi` looked a little odd (less readable) because many blank lines were removed. However, this is expected, and it follows the typeshed conventions: https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#conventions

### CODE REVIEW

**Comment by bdice** [comment]
> Looks pretty neat except for one nit.
> 
> We might also need to upgrade version in CI: [rapidsai/integration#74](https://github.com/rapidsai/integration/pull/74)
> 
> cc: @ajschmidt8 for it, or can `cudf` independently live with a local upgrade in CI?

cuDF is now independent of the versions specified in the integration repo, because CI uses pre-commit for its linting since PR #9412. Pre-commit installs the versions specified in `.pre-commit-config.yaml` in its own virtual environment cache, which is outside of the user's conda or system environments.

**Comment by bdice** [comment]
> > > We might also need to upgrade version in CI: [rapidsai/integration#74](https://github.com/rapidsai/integration/pull/74)
> > > cc: @ajschmidt8 for it, or can `cudf` independently live with a local upgrade in CI?
> > 
> > 
> > Good catch @galipremsagar. Ideally, we would sync these versions.
> 
> To clarify, the changes in #9412 to use pre-commit _prevent_ the kind of conflicts that were solved in [rapidsai/integration#74](https://github.com/rapidsai/integration/pull/74). The versions of black/isort/flake8 are no longer specified in the cudf conda environments and installation/execution of style checkers is handled by pre-commit. The linter script does not call the linters in the conda/system environment, it allows pre-commit to handle it in its own virtual environment cache.
> 
> To the best of my knowledge, changing the versions of style checkers in the integration repo will no longer affect cudf at all. Of course, updating regularly is a good idea, but it will only have effects on non-cudf repositories in my understanding. The effect of changing the integration repo's pinnings _may_ cause conflicts with other RAPIDS repositories.
> 
> Moving to this pre-commit managed solution allows us to update more frequently and fearlessly because it decreases (or removes) the possibility of conflict with style checker versions pinned by other libraries and ensures that user environments (if managed by pre-commit) have style checkers that are always in sync with the repository and the versions used by CI.
> 
> Since `black`/`flake8`/`isort` were removed from the cudf conda environment, developers should trigger all style checkers via `pre-commit run` (or by using a git hook via `pre-commit install`) -- this is automatically managed and always in sync with the repository configuration and CI. Unlike with the conda solution, users don't have to recreate the environment to update style checkers. Pulling a new pre-commit config will automatically handle all style checker updates on the next invocation of pre-commit. Users who prefer to install and run each style checker manually/individually are on their own for compliance with CI & version updates.

@galipremsagar @ajschmidt8 After some further investigation and discussion with @vyasr, I need to retract some of this and roll back some changes. There is a specific conflict with VSCode and rapids-compose that would affect many cuDF developers. This affects part of #9412, part of #9433, and all of #9448 / #9450.

tl;dr This specific recommendation was invalid:
> developers should trigger all style checkers via `pre-commit run` (or by using a git hook via `pre-commit install`)

because it conflicts with VSCode IDE integration, which uses the conda environment.

Many of our developers use VSCode and rapids-compose. This uses the Python tools (linters, formatters, type checkers) installed in the active conda environment (the rapids-compose conda environment). This means that if the conda environment specifies `black==19.10b0`, then VSCode will use the conda env's version for "format on save" or "format document" commands. This leads to conflicts in formatting with the pinning in pre-commit's config. The technical limitation that I did not foresee is that it does not appear that VSCode can be configured to execute `pre-commit run black` (which would use the pre-commit pinning) instead of `black` (which uses the conda env). It is my guess that this would affect mypy type checking and other tools as well but I haven't verified that.

If we want to upgrade linters, we _do_ need to go through the integration repo and apply changes to all RAPIDS repos at the same time as @galipremsagar indicated. For now, I am going to close this PR and the mypy PR (#9448), and open a new PR that reverts the version changes in #9433 and re-adds the linter pinnings that @vyasr removed in #9412.

It's not an ideal situation to be pinned at two year old versions of our linting tools, but it will require a committed effort across the team to change those pinnings all together. We will continue to need two sources of truth (the conda environment and pre-commit config) that share version pinnings, because there is no other way I know of that can support the variety of configurations (pre-commit / git hooks executing outside of a compose container or conda environment, VSCode support, consistency with CI versions, etc.) that we currently support.
